graphql_compiler/query_pagination/pagination_planning.py

```diff
@@ -36,18 +46,48 @@ def try_get_pagination_plan(schema_info, query_ast, number_of_pages):
 
     pagination_field = schema_info.pagination_keys.get(pagination_node.name.value)
     if pagination_field is None:
-        return None
+        raise GraphQLPaginationError(
+            u"Cannot paginate because no pagination field is specified "
+            u"on the query root. The pagination planner is not good "
+            u"enough to consider other vertices."
+        )
 
     if is_uuid4_type(schema_info, pagination_node.name.value, pagination_field):
-        pass
-    elif is_int_field_type(schema_info, pagination_node.name.value, pagination_field):
-        quantiles = schema_info.pagination_keys.statistics.get_field_quantiles(
-            pagination_node, pagination_field
+        pass  # Uniformly distributed UUID fields have an infinite pagination capacity.
+    elif field_supports_range_reasoning(schema_info, pagination_node.name.value, pagination_field):
+        quantiles = schema_info.statistics.get_field_quantiles(
+            pagination_node.name.value, pagination_field
         )
         # We make sure there's more than enough quantiles because we don't interpolate.
-        if quantiles is None or len(quantiles) < 10 * number_of_pages:
-            return None
+        if quantiles is None:
+            raise GraphQLPaginationError(
+                u"Cannot paginate because no quantiles exist for {}.{}".format(
+                    pagination_node.name.value, pagination_field
+                )
+            )
+        if len(quantiles) < QUANTILE_REQUIREMENT_FACTOR * number_of_pages:
```

Comment:
I'm not sure about this. On one hand, we say we can't paginate — which is fine. On the other, when that happens, people are still going to want to run the query, and they'll probably run it without pagination.

It seems to me like it might be better to make some sort of interpolated best-effort guess and paginate anyway, rather than refusing to paginate and causing a more serious issue in the big picture.

Reply:
Interpolation takes work and can be avoided by just getting more stats, so I've avoided it so far.

Interpolation implies the cost-estimator should interpolate too. I haven't had the time to see if that's the best idea.

However, you bring up a good point, and the least I can do is set `QUANTILE_REQUIREMENT_FACTOR` to 1, or just remove it.

Reply:
Also, if there's 10 quantiles but the user wants 100 pages, I can also say "I can't do 100, but here's 10 pages". That's easy to do.

Reply:
Also if we need 100 pages on a query that looks like `A -> B`, and the ideal pagination location is `A` but it can support only 10 pages, we may want to produce a plan like "10 pages at `A`, and for each of them, 10 pages at `B`". While this is not the ideal plan, it's better than not paginating at all, and arguably it's better (up to a point) than only producing 10 pages (the ones at `A`). Where it ceases to be better than only producing 10 pages at `A` is if we actually needed say 10000 pages total, and we needed to split `B` into so many pieces that we suspect our pages might be severely unbalanced due to `A-B` correlations, but that scenario ends badly no matter what we do.

Obviously, if we are ever in this situation, the correct long-term answer is to collect more stats at `A`. This discussion is only about the fall-back scenario if we haven't yet collected all the stats we want.

Reply:
Yes. But plans with multiple vertex partitions are complicated because they require a change of API. They return 3 continuation queries. Maybe we should change the api to return a list of continuations, while it's still early.

Reply:
I made it return the best plan it can find, and also return some warnings. It's up to the caller to decide what to do with those warnings. I think it would make sense to use them as auxiliary information in the slow query log (or badly paginated query log), to inform what action to take.